---
title: "danRerLib: a python package for zebrafish transcriptomics"
collection: publications
# permalink: /publication/2021-05-07-mathematical-modeling-of-the-interaction-between-yolk-utilization-and-fish-growth-in-zebrafish-danio-rerio
# excerpt: 'This paper is about the number 2. The number 3 is left for future work.'
date: 2024-05-06
venue: 'Bioinformatics Advances'
link: 'https://doi.org/10.1093/bioadv/vbae065'
# paperurl: 'http://academicpages.github.io/files/paper2.pdf'
citation: 'Schwartz, A.V.; Sant, K.E.; George, U.Z. danRerLib: a python package for zebrafish transcriptomics. Bioinformatics Advances. 2024, vbae065.'
abstract: "Understanding the pathways and biological processes underlying differential gene expression is fundamental for characterizing gene expression changes in response to an experimental condition. Zebrafish, with a transcriptome closely mirroring that of humans, are frequently utilized as a model for human development and disease. However, a challenge arises due to the incomplete annotations of zebrafish pathways and biological processes, with more comprehensive annotations existing in humans. This incompleteness may result in biased functional enrichment findings and loss of knowledge. danRerLib, a versatile Python package for zebrafish transcriptomics researchers, overcomes this challenge and provides a suite of tools to be executed in Python including gene ID mapping, orthology mapping for the zebrafish and human taxonomy, and functional enrichment analysis utilizing the latest updated Gene Ontology (GO) and Kyoto Encyclopedia of Genes and Genomes (KEGG) databases. danRerLib enables functional enrichment analysis for GO and KEGG pathways, even when they lack direct zebrafish annotations through the orthology of human-annotated functional annotations. This approach enables researchers to extend their analysis to a wider range of pathways, elucidating additional mechanisms of interest and greater insight into experimental results."
---
